modules/assim.sequential/R/SDA_OBS_Assembler.R

```diff
@@ -0,0 +1,175 @@
+#' Assembler for preparing obs.mean and obs.cov for the SDA workflow
+#'
+#' @param settings_dir the path of settings.xml object.
+#' @param Var Variable name, currently support: SMP, AGB, and LAI.
+#' @param OutDir the path to store obs.mean and obs.cov
+#' @param Obs_Prep if your settings object doesn't contain Obs_Prep, you can import it separately (details see L17-18).
+#' @param skip_buffer flag to skip calculating min var based on buffer area for agb data.
+#'
+#' @return list of obs.mean and obs.cov
+#' @export
+#' @author Dongchen Zhang
+#'
+#' @examples
+#' \dontrun{
+#' settings_dir <- "/projectnb/dietzelab/dongchen/All_NEON_SDA/NEON42/SDA/pecan.xml"
+#' OutDir <- "/projectnb/dietzelab/dongchen/All_NEON_SDA/test_OBS"
+#' Var <- c("SMP", "LAI", "AGB")
+#' OBS <- SDA_OBS_Assembler(settings_dir, Var, OutDir)
+#' }
+
+
+SDA_OBS_Assembler <- function(settings_dir, Var, OutDir, Obs_Prep = NULL, skip_buffer = TRUE){
+  #export special operator
+  `%>%` <- magrittr::`%>%` 
+  `%m+%` <- as.function(lubridate::`%m+%`)
+  
+  #read settings
+  settings <- PEcAn.settings::read.settings(settings_dir)
+  
+  # I prefer to include Obs_Prep info in the settings object, that will make life easier!
+  # if you don't like it, you can provide the Obs_Prep object 
+  # followed strictly by L76 - L82 in "/pecan/modules/assim.sequential/inst/MultiSite-Exs/SDA/Create_Multi_settings.R"
+  if(!is.null(settings$state.data.assimilation$Obs_Prep)){
+    Obs_Prep <- settings$state.data.assimilation$Obs_Prep
+  }else if(is.null(Obs_Prep)){
+    PEcAn.logger::logger.info("Please provide info of Obs_Prep object!")
+    return(0)
+  }
+  
+  #prepare site_info offline, because we need to submit this to GEO, which doesn't support Bety connection.
+  
+  Site_Info <- list(site_id = settings %>% purrr::map(~.x[['run']] ) %>% purrr::map('site') %>% purrr::map('id') %>% unlist() %>% as.character(),
+                    lat = settings %>% purrr::map(~.x[['run']] ) %>% purrr::map('site') %>% purrr::map('lat') %>% unlist() %>% as.numeric(),
+                    lon = settings %>% purrr::map(~.x[['run']] ) %>% purrr::map('site') %>% purrr::map('lon') %>% unlist() %>% as.numeric(),
+                    site_name = rep("name", length(settings %>% purrr::map(~.x[['run']] ) %>% purrr::map('site') %>% purrr::map('lat') %>% unlist() %>% as.numeric())))
+  
+  #collect time points
+  #we need to know which var we want to proceed 
+  #cause for every variable we assigned the same time_step object, we only need to grab it from any of what we have here.
+  #here we grab the first var to calculate the time step.
```

Comment:
This seems like a really strong design assumption that wasn't in the overall pecan documentation or the overall function documentation. More to the point, I don't this is the right thing to do -- to me it seems like the timestep of the SDA should be a high-level argument within `<state.data.assimilation>`, not something inferred based on the order the variables are listed in the obs_prep.  Furthermore that timestep should be specified in a way that is more well defined than what you have so far (e.g. saying that a variable is annual tells you noting about WHEN in the year it represents). I suspect that as I dive deeper into this PR I'm going to find a lot of hard-coded July dates that should be settings and/or function arguments.

Reply:
Additional point, what you're doing here will also fail if the latter time points are coarser than the specified one. For example, if the assimilation timestep is monthly, you don't want to overwrite the AGB timestep to now be monthly when the data itself is annual. The obs_mean and obs_cov should be monthly but the AGB data should only show up in obs_mean and obs_cov once a year, and be absent (not NA) the rest of the year

Reply:
looking back at the pecan.xml, why aren't you just using the `<forecast.time.step>` variable?

Reply:
1. I changed the design such that the time points object will only be calculated once within the assembler function and then be passed to the rest of the prep functions.
2. The `<forecast.time.step>` won't work because it only provides the unit but does not include the number we set currently.

Reply:
And to reply to the first point, there aren't any hard-coded something July at all for all the functions that I created. Instead, the time points are calculated only based on the information from the settings object.